### ISSUE TITLE
Failed to initialize DPDK on EC2 i3 instance

### ISSUE THREAD

**Issue by wangli1426** [spdk/spdk#210]
Hi guys,

I am a database guy who are trying to use the NVMe. On my EC2 i3 instance (Ubuntu 16.04), I make sure that the instance is equipped with nvm by running `sudo nvme list`. Then I compile SPDK and run `scripts/setup.sh`. However, when I run the hello world example, I get the following information. 

```Starting DPDK 17.08.0 initialization...
[ DPDK EAL parameters: hello_world -c 0x1 --file-prefix=spdk0 --base-virtaddr=0x1000000000 --proc-type=auto ]
EAL: Detected 2 lcore(s)
EAL: Auto-detected process type: PRIMARY
EAL: No free hugepages reported in hugepages-1048576kB
EAL: No free hugepages reported in hugepages-2048kB
EAL: FATAL: Cannot get hugepage information.
EAL: Cannot get hugepage information.
Failed to initialize DPDK
```
Any help is appreciated. 

Also, How can reverse the  unbound operations in scripts/setup.sh?

**Comment by optimistyzy** [comment]
./scripts/setup.sh reset 
 BTW, is there really having hugepages allocated on you EC2 i3 instance?

**Comment by wangli1426** [comment]
Thanks for your response.
I re-run `sudo /scripts/setup.sh` command and it shows:

```modprobe: FATAL: Module uio_pci_generic not found in directory /lib/modules/4.4.0-1022-aws
0000:00:1e.0 (1d0f cd01): nvme -> uio_pci_generic
scripts/setup.sh: line 37: /sys/bus/pci/drivers/uio_pci_generic/new_id: No such file or directory
scripts/setup.sh: line 38: /sys/bus/pci/drivers/uio_pci_generic/bind: No such file or directory
```

Any suggestions?

**Comment by wangli1426** [comment]
Also, the hugepage is enabled.
```
# sudo cat /sys/kernel/mm/transparent_hugepage/enabled
[always] madvise never
```

**Comment by jimharris** [comment]
> On Oct 16, 2017, at 8:01 AM, Li Wang <notifications@github.com> wrote:
> 
> Thanks for your response.
> I re-run sudo /scripts/setup.sh command and it shows:
> 
> 0000:00:1e.0 (1d0f cd01): nvme -> uio_pci_generic
> scripts/setup.sh: line 37: /sys/bus/pci/drivers/uio_pci_generic/new_id: No such file or directory
> scripts/setup.sh: line 38: /sys/bus/pci/drivers/uio_pci_generic/bind: No such file or directory
> 
> Any suggestions?

Hi Li,

Can you check if the uio_pci_generic module is available on your EC2 instance?

Also if you could post your /proc/meminfo that may provide some hints on the hugepage errors reported when starting up the hello_world app.

Thanks,

-Jim

**Comment by anakli** [comment]
Were you able to resolve this issue? I am seeing a similar problem trying to run setup.sh from SPDK 17.07 an EC2 i3 instance (see output below). It seems like the uio_pci_generic module is not available. 

I was previously using a version of SPDK from Aug 2016 (commit  d7120a3ee88062e06049e195131c703522e1dfda) and this worked fine on EC2 -- I believe this is because SPDK was using igb_uio instead. 

Any suggestions for running SPDK 17.07 on EC2? Thanks!

`modprobe: FATAL: Module uio_pci_generic not found in directory /lib/modules/4.4.0-1028-aws`
`0000:00:1e.0 (1d0f cd01): no driver -> uio_pci_generic`
`deps/spdk/scripts/setup.sh: line 33: /sys/bus/pci/drivers/uio_pci_generic/new_id: No such file or directory`
`deps/spdk/scripts/setup.sh: line 34: /sys/bus/pci/drivers/uio_pci_generic/bind: No such file or directory`

**Comment by anakli** [comment]
Update: Amazon Linux AMI includes uio_pci_generic, so SPDK v17.07 works when I use Amazon Linux as the image for my EC2 instance. I was previously using the Ubuntu 16.04 AMI which did not have uio_pci_generic.

**Comment by benlwalker** [comment]
The instance must have both uio_pci_generic and have the ability to allocate hugepages through a hugetlb mount point. That's different than having transparent hugepages enabled.

**Comment by rajasekaranmam** [comment]
Am facing a similar issue in Ubuntu 18.04 while running ./setup command before starting SPDK;
modprobe: FATAL: Module uio_pci_generic not found in directory /lib/modules/4.15.0-36-generic
0000:83:00.0 (8086 0953): nvme -> uio_pci_generic
./setup.sh: line 106: /sys/bus/pci/drivers/uio_pci_generic/new_id: No such file or directory
./setup.sh: line 107: /sys/bus/pci/drivers/uio_pci_generic/bind: No such file or directory
Any leads would be appreciate

**Comment by jimharris** [comment]
Hi @rajasekaranmam,

Are there any linux-image-extra packages available?  These typically have the uio_pci_generic driver in them.

Another longer term option is to modify setup.sh to use the DPDK igb_uio driver if it is available.  SPDK has recently added a config option to build igb_uio since it is needed in some cases when using QuickAssist hardware for crypto offload.  setup.sh could detect when uio_pci_generic is not available and look for igb_uio instead.  If it's not available, it could direct the user to ./configure --with-igb-uio-driver and build and install.

-Jim

**Comment by jimharris** [comment]
Closing this issue.  linux-image-extra packages should add uio_pci_generic.  If there are cases where this does not happen, please file a new issue.  Thanks.
